src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.cs

```diff
@@ -603,15 +603,32 @@ internal PragmaWarningState GetPragmaDirectiveWarningState(string id, int positi
             if (_lazyNullableDirectiveMap == null)
             {
                 // Create the #nullable directive map on demand.
-                Interlocked.CompareExchange(ref _lazyNullableDirectiveMap, NullableDirectiveMap.Create(this), null);
+                Interlocked.CompareExchange(ref _lazyNullableDirectiveMap, NullableDirectiveMap.Create(this, IsGeneratedCode()), null);
             }
 
             return _lazyNullableDirectiveMap.GetDirectiveState(position);
         }
 
+        internal bool IsGeneratedCode()
+        {
+            if (_lazyIsGeneratedCode == -1)
+            {
+                // Create the generated code status on demand
+                bool isGenerated = GeneratedCodeUtilities.IsGeneratedCode(
+                           this,
+                           isComment: trivia => trivia.Kind() == SyntaxKind.SingleLineCommentTrivia || trivia.Kind() == SyntaxKind.MultiLineCommentTrivia,
+                           cancellationToken: default);
+
+                Interlocked.CompareExchange(ref _lazyIsGeneratedCode, isGenerated ? 1 : 0, -1);
```

Comment:
>Interlocked.CompareExchange [](start = 16, length = 27)

`Interlocked.CompareExchange()` shouldn't be necessary for an `int`.

Reply:
Oh, yeah, that makes sense. We've already done the work anyway, and we don't care which thread sets it. That actually allows this code to be much simpler.